### DISCUSSION TITLE
Avoiding DOS via recursive explosion?

### DISCUSSION THREAD

**Discussion by nielsbasjes** [graphql-java/graphql-java#2975] [Q&A] [answered]
Let's assume I have a tree structure I want to expose in GraphQL and I create a schema that is (simplified):
```
type Query {
  root: Node
}

type Node {
  name: String
  parent: Node
  children: [Node]
}
```

Now even with a relatively small tree (root node with 100 children) a malicious client can do a lookup like this:
```
query {
  root {
    children {
      parent {
        children {
          parent {
            children {
              parent {
                name
              }
            }
          }
        }
      }
    }
  }
}
```

This will in general make the server instance crash with something like an out of memory or stack overflow kind of problem.

My current ideas to avoid this are about setting a request scoped counter that limits the levels deep your query can be.

Are there better ways to mitigate this problem?
Are there features in GraphQL-Java (or Spring-GraphQL which I use right now) that can support in handling this?
What kind of mitigation do you use/advise?

**Comment by bbakerman** [comment, marked as answer]
See `graphql.analysis.MaxQueryDepthInstrumentation` and `graphql.analysis.MaxQueryComplexityInstrumentation`

I also recommend a max query size to how much input you are willing to accept.

**Reply by nielsbasjes** [reply]
Thanks, these two work great.
You mention limiting the max query size, is that the MaxQueryComplexityInstrumentation or do you mean something different?

**Reply by nielsbasjes** [reply]
For future reference for others finding this discussion.
I found that if the limits needed for your application are low then the `IntrospectionQuery` used by GraphiQL to fetch the schema fails because it is also "too complex and too deep".

My current workaround is to add a filter on these Instrumentations to skip their checks if this is only the retrieval of the schema.

My current code looks like this (I wrote this, I put this under Apache 2/MIT/BSD/... license: Have fun with it and don't come to me if it causes problems):
```

    /*
    * When retrieving a field we need to avoid getting too much data.
    * Yet when retrieving the schema we need to allow a much bigger query.
    */
    @SuppressWarnings("rawtypes")
    private static boolean isOnlySchemaQuery(InstrumentationExecuteOperationParameters parameters) {
        OperationDefinition operationDefinition = parameters.getExecutionContext().getOperationDefinition();
        List<Selection> selections = operationDefinition.getSelectionSet().getSelections();
        if (selections.size() != 1) {
            return false;
        }
        Selection selection = selections.get(0);
        if (!(selection instanceof Field)) {
            return false;
        }
        return ((Field) selection).getName().equals("__schema");
    }

    public static MaxQueryDepthInstrumentation getMaxDepthInstrumentation(int maxDepth) {
        return new MaxQueryDepthInstrumentation(maxDepth) {
            @Override
            public InstrumentationContext<ExecutionResult>
                    beginExecuteOperation(
                        InstrumentationExecuteOperationParameters parameters,
                        InstrumentationState state) {
                if (isOnlySchemaQuery(parameters)) {
                    return noOp();
                }
                return super.beginExecuteOperation(parameters, state);
            }
        };
    }

    public static MaxQueryComplexityInstrumentation getMaxComplexityInstrumentation(int maxComplexity) {
        return new MaxQueryComplexityInstrumentation(maxComplexity) {
            @Override
            public InstrumentationContext<ExecutionResult>
                    beginExecuteOperation(
                        InstrumentationExecuteOperationParameters parameters,
                        InstrumentationState state) {
                if (isOnlySchemaQuery(parameters)) {
                    return noOp();
                }
                return super.beginExecuteOperation(parameters, state);
            }
        };
    }
```

which can then be used with something like this:

```
    @Bean
    GraphQlSourceBuilderCustomizer graphQlSourceBuilderCustomizer() {
        return builder -> builder
            // Reduce the possibilities of sending the server into an OOM error.
            .instrumentation(List.of(
                getMaxDepthInstrumentation(6),
                getMaxComplexityInstrumentation(30)
            ))
            .schemaFactory(
...
```

I'm not sure how generic this problem is. 
If you guys think it is generic then please reuse/copy the above code/ideas into the project.

**Comment by bbakerman** [comment]
> You mention limiting the max query size, is that the MaxQueryComplexityInstrumentation

`MaxQueryComplexityInstrumentation` allows you to assign weights to fields and then have that be calculated for you.

I mostly just meant at a transport level, have a maximum amount if input you will accept - so that consumers cant throw 100MB queries at you - even if they wont parse or execute - they will consume memory initially.

graphql-java has a `graphql.parser.ParserOptions#MAX_QUERY_TOKENS` that restricts how many tokens it will allow BUT if you allow unlimited input to start with, this will be undermined
